element-view/src/androidMain/kotlin/ElementView.kt

```diff
@@ -0,0 +1,176 @@
+        val bitmap = adapter?.bitmap
+        if (bitmap != null) {
+            canvas.drawBitmap(bitmap, 0f, 0f, blitPaint)
+        }
+    }
+
+    public class Adapter<T>(
+        private val dataSource: Flow<T>,
+        private val updater: Updater<T>,
+    ) {
+
+        /** Android's Size was added in API 21, so do this to avoid bumping min version. */
+        public data class State(val view: ElementView?, val root: RootElement, val width: Int, val height: Int)
+
+        public fun interface Updater<T> {
+            public fun update(root: RootElement, width: Float, height: Float, data: T)
+        }
+
+        private val state: MutableStateFlow<State> = MutableStateFlow(EMPTY_STATE)
+
+        /** Coroutine scope rendering occurs on. This scope is live between calls to [onAttached] and [onDetached]. */
+        private var renderScope: CoroutineScope? = null
+            set(value) {
+                if (field !== value) field?.cancel()
+                field = value
+            }
+
+        /** The most recent bitmap to finish rendering, if any. */
+        internal var bitmap: Bitmap? = null
+            private set
+
+        /**
+         * The view changed size, so we should re-render the bitmap. When this happens, the RootElement
+         * will be reset, so the render will have a clean slate.
+         */
+        internal fun onSizeChanged(width: Int, height: Int) {
+            state.value = state.value.copy(root = RootElement(), width = width, height = height)
+        }
+
+        /**
+         * Attach this to a view (which is itself attached to a window).
+         * This will immediately render the current state if possible.
+         */
+        internal fun onAttached(view: ElementView) {
+            state.value = State(view, RootElement(), view.width, view.height)
+            renderScope = object : CoroutineScope {
+                private val job = Job()
+                override val coroutineContext: CoroutineContext get() = job
+            }.also { launchRenderingIn(it) }
+        }
+
+        /** Detach this from a view (usually because the view was detached from a window). */
+        internal fun onDetached() {
+            renderScope = null
+            state.value = EMPTY_STATE
+        }
+
+        private fun launchRenderingIn(scope: CoroutineScope) = scope.launch {
+            // Unlimited buffer looks scary, but collection has way higher throughput than emission, so this doesn't grow large.
+            val buffer = MutableSharedFlow<Bitmap>(extraBufferCapacity = Int.MAX_VALUE)
+            val pool = BitmapPool()
+            scope.launch(Dispatchers.Main.immediate) {
+                buffer.collect { bitmap ->
+                    this@Adapter.bitmap?.let { pool.release(it) }
+                    this@Adapter.bitmap = bitmap
+                    state.value.view?.invalidate()
+                }
+            }
+            state.collectLatest { state ->
```

Comment:
nit: maybe a little easier to read if you moved the function's `scope.launch` to only wrap this `state.collectLatest`?
At first I though you intended to have line 127's `scope.launch` be a child of the function's `scope.launch`; but I think you just wanted to parallel running Coroutines that don't need a parent-child relationship. So, having two `scope.launch`'s next to each other (at the same indentation) makes to intention easier to see. 🤷